Signal/src/ViewControllers/OWSViewController.m

```diff
@@ -3,15 +3,128 @@
 //
 
 #import "OWSViewController.h"
+#import "UIView+OWS.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
+@interface OWSViewController ()
+
+@property (nonatomic, weak) UIView *bottomLayoutView;
+@property (nonatomic) NSLayoutConstraint *bottomLayoutConstraint;
+
+@end
+
+#pragma mark -
+
 @implementation OWSViewController
 
 - (void)dealloc
 {
     // Surface memory leaks by logging the deallocation of view controllers.
     DDLogVerbose(@"Dealloc: %@", self.class);
+
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)autoPinViewToBottomGuideOrKeyboard:(UIView *)view
+{
+    OWSAssert(view);
+    OWSAssert(!self.bottomLayoutConstraint);
+
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(keyboardWillShow:)
+                                                 name:UIKeyboardWillShowNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(keyboardDidShow:)
+                                                 name:UIKeyboardDidShowNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(keyboardWillHide:)
+                                                 name:UIKeyboardWillHideNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(keyboardDidHide:)
+                                                 name:UIKeyboardDidHideNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(keyboardWillChangeFrame:)
+                                                 name:UIKeyboardWillChangeFrameNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(keyboardDidChangeFrame:)
+                                                 name:UIKeyboardDidChangeFrameNotification
+                                               object:nil];
+
+    self.bottomLayoutView = view;
+    self.bottomLayoutConstraint = [view autoPinToBottomLayoutGuideOfViewController:self withInset:0];
+}
+
+- (void)keyboardWillShow:(NSNotification *)notification
```

Comment:
I could simplify this, but I'm waiting to see if we need to do anything different for each kind of event.